### PR TITLE
Add more options to expense approval by author policy

### PR DESCRIPTION
**Pull request by hdiniz**
Resolves https://github.com/opencollective/opencollective/issues/6075

Adds more options to EXPENSE_AUTHOR_CANNOT_APPROVE policy so that it can be gradually adopted by hosts that want to enforce it to hosted collectives.

### CODE REVIEW

File: server/graphql/common/expenses.ts [hunk 1]
```diff
@@ -382,14 +382,38 @@ export const canApprove: ExpensePermissionEvaluator = async (req, expense, optio
...
+
+    const policy = hostPolicy.enabled && hostPolicy.appliesToHostedCollectives ? hostPolicy : collectivePolicy;
+    const currency = hostPolicy.enabled ? expense.collective.host.currency : expense.collective.currency;
+
+    if (policy.enabled && expense.amount >= policy.amountInCents) {
+      const collectiveAdminCount = await req.loaders.Member.countAdminMembersOfCollective.load(expense.collective.id);
+
+      const shouldEnforcePolicy =
+        collectiveAdminCount >= 2 ||
+        (collectivePolicy.enabled && !hostPolicy.enabled) ||
```
Comment (kewitz) [inline server/graphql/common/expenses.ts:399]: Having this inside a conditional for `policy.enabled` after evaluating that from `collectivePolicy` and `hostPolicy` looks a bit convoluted. Could you add a comment on what exactly we're testing here?

```ts
        (collectivePolicy.enabled && !hostPolicy.enabled) ||
```

File: server/graphql/v2/object/Permission.ts [hunk 2]
```diff
@@ -8,6 +9,7 @@ export const Permission = new GraphQLObjectType({
   fields: () => ({
     allowed: { type: new GraphQLNonNull(GraphQLBoolean) },
     reason: { type: GraphQLString },
+    reasonDetails: { type: GraphQLJSON },
```
Comment (kewitz) [inline server/graphql/v2/object/Permission.ts:12]: Nice!

File: server/graphql/loaders/members.ts [hunk 3]
```diff
@@ -54,6 +54,21 @@ export const generateAdminUsersEmailsForCollectiveLoader = () => {
   );
 };
 
+export const generateCountAdminMembersOfCollective = () => {
```
Comment (kewitz) [inline server/graphql/loaders/members.ts:57]: 👍🏼

File: migrations/20221212200839-migrate-author-cannot-approve-expense.js [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+                  'appliesToHostedCollectives',
+                  false,
+                  'appliesToSingleAdminCollectives',
+                  false
+                )
+              )
+          )
+      WHERE data#>'{policies,EXPENSE_AUTHOR_CANNOT_APPROVE}' is not null;
+    `);
+  },
```
Comment (kewitz) [inline migrations/20221212200839-migrate-author-cannot-approve-expense.js:29]: Looks good, make sure you test with a snapshot and that it works as expected on staging.

**Review by kewitz — APPROVED** [review]

File: server/graphql/common/expenses.ts (hunk 1 above)
Comment (hdiniz) [reply server/graphql/common/expenses.ts]: I agree 100%, I tried to refactor it a bit now. let me know what you think.

File: server/graphql/common/expenses.ts [hunk 5]
```diff
@@ -382,11 +382,36 @@ export const canApprove: ExpensePermissionEvaluator = async (req, expense, optio
     return false;
   } else {
     expense.collective = expense.collective || (await req.loaders.Collective.byId.load(expense.CollectiveId));
-    if (hasPolicy(expense.collective, POLICIES.EXPENSE_AUTHOR_CANNOT_APPROVE) && req.remoteUser.id === expense.UserId) {
+    expense.collective.host =
+      expense.collective.host || (await req.loaders.Collective.byId.load(expense.collective.HostCollectiveId));
```
Comment (Betree) [inline server/graphql/common/expenses.ts:386]: @hdiniz it is valid for a collective to receive expenses without being connected to a fiscal host yet, we should ignore if the `HostCollectiveId` is null, or if `approvedAt` is null. I believe this is causing the following error: https://sentry.io/organizations/open-collective/issues/3814083338/?project=5199682

**Pull request merged**